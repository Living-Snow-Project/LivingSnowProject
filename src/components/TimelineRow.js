import React from "react";
import { Pressable, StyleSheet, Text, View } from "react-native";
import PropTypes from "prop-types";
import { PictureIcon, RecordIcon } from "../components/TabBarIcon";
import { Routes } from "../navigation/Routes";

function empty(text) {
  return !text;
}

function topText({ date, name, organization, latitude, longitude }) {
  let result = `${date.slice(0, 10)}\n`;

  if (!empty(name)) {
    result += name;

    if (!empty(organization)) {
      result += ` (${organization})`;
    }
  } else {
    result += `unknown Scientist`;
  }

  result += `\nLocation: ${latitude}, ${longitude}`;

  return result;
}

function bottomText({ locationDescription, notes }) {
  let result = "";
  let newline = "";

  if (!empty(locationDescription)) {
    result += `Description: ${locationDescription}`;
    newline = "\n";
  }

  if (!empty(notes)) {
    result += `${newline}Notes: ${notes}`;
  }

  return result;
}

// TODO: make a recordClosure object
function parsePhotoUris(photoUris) {
  if (!photoUris) {
    return [];
  }

  let result = photoUris.split(`;`);
  result.pop();

  return result;
}

const TimelineRow = ({ navigation, record, showAll = false }) => {
  const isAtlas = record.type.includes(`Atlas`);

  // this logic should be in a parent function
  if (
    !showAll &&
    ((!global.appConfig.showAtlasRecords && isAtlas) ||
      (global.appConfig.showOnlyAtlasRecords && !isAtlas))
  ) {
    return null;
  }

  return (
    <View style={styles.recordContainer}>
      <Pressable
        onPress={() =>
          navigation.navigate(Routes.RecordDetailsScreen, { record: record })
        }
      >
        <View style={styles.recordTop}>
          <View style={styles.topText}>
            <Text>{topText(record)}</Text>
          </View>
          <View style={styles.topIcon}>
            <RecordIcon type={record.type} />
          </View>
          {!empty(record.photoUris) && (
            <View style={styles.topIcon}>
              <PictureIcon />
              <View style={styles.photosNotification}>
                <Text style={{ color: "white" }}>
                  {parsePhotoUris(record.photoUris).length}
                </Text>
              </View>
            </View>
          )}
        </View>
        {(!empty(record.locationDescription) || !empty(record.notes)) && (
          <Text>{bottomText(record)}</Text>
        )}
      </Pressable>
    </View>
  );
};

TimelineRow.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,
  }),
  record: PropTypes.shape({
    type: PropTypes.string.isRequired,
    photoUris: PropTypes.string,
    locationDescription: PropTypes.string,
    notes: PropTypes.string,
  }),
  showAll: PropTypes.bool,
};

export { TimelineRow };

const styles = StyleSheet.create({
  recordContainer: {
    borderBottomWidth: 1,
  },
  recordTop: {
    flexDirection: "row",
    flex: 1,
  },
  topText: {
    marginLeft: 1,
    flex: 0.7,
  },
  topIcon: {
    marginLeft: 1,
    flex: 0.15,
  },
  photosNotification: {
    position: "absolute",
    alignSelf: "flex-end",
    top: 1,
    right: 2,
    backgroundColor: "red",
    borderRadius: 40,
    alignItems: "center",
    width: "33%",
  },
});
