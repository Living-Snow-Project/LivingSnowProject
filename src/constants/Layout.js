import { Dimensions } from "react-native";

const { height, width } = Dimensions.get("window");

export default {
  window: {
    width,
    height,
  },
  isSmallDevice: width < 375,
};
