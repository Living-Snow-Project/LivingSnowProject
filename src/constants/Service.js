export const serviceEndpoint = 'https://snowalgaeproductionapp.azurewebsites.net';
