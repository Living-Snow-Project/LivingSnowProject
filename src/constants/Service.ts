const serviceEndpoint = "https://snowalgaeproductionapp.azurewebsites.net";
const photosBlobStorageEndpoint =
  "https://snowalgaestorage.blob.core.windows.net/photos";

export { serviceEndpoint, photosBlobStorageEndpoint };
