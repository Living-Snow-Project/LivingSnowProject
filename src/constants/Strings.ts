// for TextInput placeholder prop
const Placeholders = {
  GPS: {
    AcquiringLocation: "Looking for GPS signal",
    EnterCoordinates: "ie. 12.345678, -123.456789",
    NoPermissions:
      "Check location permissions in Settings. Enter coordinates manually.",
    NoLocation: "Could not determine location. Enter coordinates manually.",
  },
  // TODO: RecordScreen placeholders
};

const Labels = {
  TimelineScreen: {
    DownloadedRecords: "Downloaded",
    NoRecords: "No records to display",
    PendingRecords: "Pending",
  },
  StatusBar: {
    NoConnection: "No Internet Connection",
  },
};

// for Alert.alert calls
const Notifications = {
  uploadSuccess: {
    title: "Upload succeeded",
    message: "Thanks for your submission.",
  },
  uploadFailed: {
    title: "Record Saved",
    message: "We will upload it later.",
  },
  invalidCoordinates: {
    title: "Invalid GPS coordinates",
    message:
      'Coordinates must be in "lat, long" format. ie. 12.345678, -123.456789',
  },
};

export { Notifications, Placeholders, Labels };
