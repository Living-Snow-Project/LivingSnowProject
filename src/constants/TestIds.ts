const TestIds = {
  GPS: {
    gpsManualInputTestId: "gps-manual-input",
    gpsManualPressableTestId: "gps-manual-pressable",
  },
  Photos: {
    photoSelectorTestId: "photo-selector",
  },
  Pickers: {
    recordSelectorTestId: "record-type-picker",
    atlastSelectorTestId: "atlas-type-picker",
  },
  RecordScreen: {
    UploadButton: "record-screen-upload-button",
  },
  TimelineScreen: {
    RefreshControl: "timeline-screen-refresh-control",
    SettingsButton: "timeline-screen-settings-button",
    NewRecordButton: "timeline-screen-record-button",
  },
  SettingsScreen: {
    ShowGpsWarning: "settings-screen-show-gps-warning",
    ShowAtlasRecords: "settings-screen-show-atlas-records",
    ShowOnlyAtlasRecords: "settings-screen-show-only-atlas-records",
  },
  Icons: {
    SampleIcon: "sample-icon",
    SightingIcon: "sighting-icon",
    PictureIcon: "picture-icon",
    DeleteIcon: "delete-icon",
    EditIcon: "edit-icon",
    SnowIcon: "snow-icon",
  },
  RecordList: {
    EditRecordAction: "edit-record-action",
    DeleteRecordAction: "delete-record-action",
  },
};

export default TestIds;
