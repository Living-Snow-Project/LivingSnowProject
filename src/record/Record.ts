import { RecordDescription } from "../constants/Strings";

type RecordTypePickerItem = {
  value: AlgaeRecordType;
  label: string;
};

// specific format for RNPickerSelect
const recordTypePickerItems: RecordTypePickerItem[] = [
  { value: "Sample", label: RecordDescription.Sample },
  { value: "Sighting", label: RecordDescription.Sighting },
  { value: "Atlas: Red Dot", label: RecordDescription.AtlasRedDot },
  {
    value: "Atlas: Red Dot with Sample",
    label: RecordDescription.AtlasRedDotWithSample,
  },
  { value: "Atlas: Blue Dot", label: RecordDescription.AtlasBlueDot },
  {
    value: "Atlas: Blue Dot with Sample",
    label: RecordDescription.AtlasBlueDotWithSample,
  },
];

const getAllRecordTypePickerItems = (): RecordTypePickerItem[] =>
  recordTypePickerItems;

const getRecordTypePickerItem = (
  type: AlgaeRecordType
): RecordTypePickerItem => {
  const result = recordTypePickerItems.find((cur) => cur.value === type);

  if (result === undefined) {
    return { value: "Undefined", label: RecordDescription.Undefined };
  }

  return result;
};

const isSample = (type: AlgaeRecordType): boolean =>
  Array<AlgaeRecordType>(
    "Sample",
    "Atlas: Red Dot with Sample",
    "Atlas: Blue Dot with Sample"
  ).includes(type);

const isAtlas = (type: AlgaeRecordType): boolean =>
  Array<AlgaeRecordType>(
    "Atlas: Red Dot",
    "Atlas: Red Dot with Sample",
    "Atlas: Blue Dot",
    "Atlas: Blue Dot with Sample"
  ).includes(type);

// TODO: makeExampleRecord should also be used to seed RecordScreen (ie. no more "no records to display")
const makeExampleRecord = (type: AlgaeRecordType): AlgaeRecord => {
  const atlasType: AtlasType = isAtlas(type) ? "Snow Algae" : "Undefined";

  return {
    id: 1234,
    type,
    name: "test name",
    date: new Date("2021-09-16T00:00:00"),
    organization: "test org",
    latitude: -123.456,
    longitude: 96.96,
    tubeId: isSample(type) ? "LAB-1337" : "",
    locationDescription: "test location",
    notes: "test notes",
    atlasType,
    photos: [
      {
        uri: "46",
        size: 100,
        width: 16,
        height: 16,
      },
      { uri: "23", size: 4096, width: 128, height: 128 },
    ],
  };
};

// needed for JSON.parse, otherwise date property will become string
const recordReviver = (key: string, value: any): any => {
  if (key === "date") {
    return new Date(value);
  }

  return value;
};

// decodes AlgaeRecord or AlgaeRecord[] JSON
function jsonToRecord<T>(json: string): T {
  return JSON.parse(json, recordReviver);
}

// want to display date in YYYY-MM-DD format
const recordDateFormat = (date: Date): string => {
  const dayNum: number = date.getDate();
  let day: string = `${dayNum}`;

  if (dayNum < 10) {
    day = `0${dayNum}`;
  }

  const monthNum: number = date.getMonth() + 1;
  let month: string = `${monthNum}`;

  if (monthNum < 10) {
    month = `0${monthNum}`;
  }

  return `${date.getFullYear()}-${month}-${day}`;
};

export {
  getRecordTypePickerItem,
  getAllRecordTypePickerItems,
  makeExampleRecord,
  isSample,
  isAtlas,
  jsonToRecord,
  recordDateFormat,
};
