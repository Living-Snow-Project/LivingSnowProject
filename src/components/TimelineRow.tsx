import React from "react";
import { Pressable, StyleSheet, Text, View } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { PictureIcon, RecordIcon } from "./Icons";
import { RootStackNavigationProp } from "../navigation/Routes";
import { AlgaeRecordPropType } from "../record/PropTypes";
import { Labels } from "../constants/Strings";
import { recordDateFormat } from "../record/Record";

const styles = StyleSheet.create({
  recordContainer: {
    backgroundColor: "white",
  },
  recordTop: {
    flexDirection: "row",
    flex: 1,
  },
  topText: {
    marginLeft: 1,
    flex: 0.7,
  },
  topIcon: {
    marginLeft: 1,
    flex: 0.15,
  },
  photosNotification: {
    position: "absolute",
    alignSelf: "flex-end",
    top: 1,
    right: 2,
    backgroundColor: "red",
    borderRadius: 40,
    alignItems: "center",
    width: "33%",
  },
});

function empty(text) {
  return !text;
}

function topText({
  date,
  name,
  organization,
  latitude,
  longitude,
}: AlgaeRecord) {
  let result = `${recordDateFormat(date)}\n`;

  if (!empty(name)) {
    result += name;

    if (!empty(organization)) {
      result += ` (${organization})`;
    }
  } else {
    result += `Anonymous Scientist`;
  }

  result += `\n${Labels.RecordFields.GPSCoordinates}: ${latitude}, ${longitude}`;

  return result;
}

function bottomText({ locationDescription, notes }: AlgaeRecord) {
  let result = "";
  let newline = "";

  if (!empty(locationDescription)) {
    result += `${Labels.RecordFields.LocationDescription}: ${locationDescription}`;
    newline = "\n";
  }

  if (!empty(notes)) {
    result += `${newline}${Labels.RecordFields.Notes}: ${notes}`;
  }

  return result;
}

type TimelineRowProps = {
  record: AlgaeRecord;
};

export default function TimelineRow({ record }: TimelineRowProps) {
  const { navigate } = useNavigation<RootStackNavigationProp>();
  return (
    <View style={styles.recordContainer}>
      <Pressable
        testID={record.id.toString()}
        onPress={() => navigate("RecordDetails", { record })}
      >
        <View style={styles.recordTop}>
          <View style={styles.topText}>
            <Text>{topText(record)}</Text>
          </View>
          <View style={styles.topIcon}>
            <RecordIcon type={record.type} />
          </View>
          {record.photos !== undefined && record.photos.length > 0 && (
            <View style={styles.topIcon}>
              <PictureIcon />
              <View style={styles.photosNotification}>
                <Text style={{ color: "white" }}>{record.photos.length}</Text>
              </View>
            </View>
          )}
        </View>
        {(!empty(record.locationDescription) || !empty(record.notes)) && (
          <Text>{bottomText(record)}</Text>
        )}
      </Pressable>
    </View>
  );
}

TimelineRow.propTypes = {
  record: AlgaeRecordPropType.isRequired,
};
