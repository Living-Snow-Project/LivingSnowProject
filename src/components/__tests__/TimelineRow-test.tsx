import React from "react";
import { render } from "@testing-library/react-native";
import TimelineRow from "../TimelineRow";
import { makeExampleRecord } from "../../record/Record";
import TestIds from "../../constants/TestIds";
import { Labels } from "../../constants/Strings";

describe("TimelineRow test suite", () => {
  let expectedRecord;

  beforeEach(() => {
    expectedRecord = makeExampleRecord("Sample");
  });

  test("renders", () => {
    const { toJSON } = render(<TimelineRow record={expectedRecord} />);

    expect(toJSON()).toMatchSnapshot();
  });

  test("sample icon", () => {
    const { queryByTestId } = render(<TimelineRow record={expectedRecord} />);

    expect(queryByTestId(TestIds.Icons.SampleIcon)).toBeTruthy();
  });

  test("sighting icon", () => {
    expectedRecord = makeExampleRecord("Sighting");

    const { queryByTestId } = render(<TimelineRow record={expectedRecord} />);

    expect(queryByTestId(TestIds.Icons.SightingIcon)).toBeTruthy();
  });

  test("missing name", () => {
    expectedRecord.name = undefined;

    const { queryByText } = render(<TimelineRow record={expectedRecord} />);

    expect(queryByText(new RegExp(Labels.RecordFields.Name))).toBeFalsy();
  });

  test("missing organization", () => {
    expectedRecord.organization = undefined;

    const { queryByText } = render(<TimelineRow record={expectedRecord} />);

    expect(
      queryByText(new RegExp(Labels.RecordFields.Organization))
    ).toBeFalsy();
  });

  test("missing location description", () => {
    expectedRecord.locationDescription = undefined;

    const { queryByText } = render(<TimelineRow record={expectedRecord} />);

    expect(
      queryByText(new RegExp(Labels.RecordFields.LocationDescription))
    ).toBeFalsy();
  });

  test("missing notes", () => {
    expectedRecord.notes = undefined;

    const { queryByText } = render(<TimelineRow record={expectedRecord} />);

    expect(queryByText(new RegExp(Labels.RecordFields.Notes))).toBeFalsy();
  });

  test("photos icon", () => {
    const { queryByTestId } = render(<TimelineRow record={expectedRecord} />);

    expect(queryByTestId(TestIds.Icons.PictureIcon)).toBeTruthy();
  });

  test("no photos icon", () => {
    expectedRecord.photoUris = undefined;

    const { queryByTestId } = render(<TimelineRow record={expectedRecord} />);

    expect(queryByTestId(TestIds.Icons.PictureIcon)).toBeFalsy();
  });
});
