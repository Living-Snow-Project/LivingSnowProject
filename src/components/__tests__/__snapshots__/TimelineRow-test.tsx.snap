// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineRow test suite renders 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 1,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="1234"
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 0.7,
            "marginLeft": 1,
          }
        }
      >
        <Text>
          2021-09-16
test name (test org)
Coordinates: -123.456, 96.96
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 0.15,
            "marginLeft": 1,
          }
        }
      >
        <Text />
      </View>
      <View
        style={
          Object {
            "flex": 0.15,
            "marginLeft": 1,
          }
        }
      >
        <Text />
        <View
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "flex-end",
              "backgroundColor": "red",
              "borderRadius": 40,
              "position": "absolute",
              "right": 2,
              "top": 1,
              "width": "33%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
              }
            }
          >
            2
          </Text>
        </View>
      </View>
    </View>
    <Text>
      Location Description: test location
Notes: test notes
    </Text>
  </View>
</View>
`;
